lib/orderbook/OrderBook.ts

```diff
@@ -1,130 +1,205 @@
+      pool.on('packet.getOrders', async (peer: Peer) => {
+        // TODO: just send supported pairs
+        const pairs = await this.getPairs();
+        const outgoingOrders: OutgoingOrder[] = [];
+
+        for (const { id } of pairs) {
+          const orders = await this.getOwnOrders(id, 0);
+          await Promise.all([
+            this.castToOutgoingOrder(orders.buyOrders, outgoingOrders),
+            this.castToOutgoingOrder(orders.sellOrders, outgoingOrders),
+          ]);
+        }
+
+        const packet = new OrdersPacket({ orders: outgoingOrders });
+
+        peer.sendPacket(packet);
+      });
     }
   }
 
-  public async init() {
-    const [pairs, peerOrders, ownOrders] = await Promise.all([
-      this.repository.getPairs(),
-      this.repository.getPeerOrders(),
-      this.repository.getOwnOrders(),
-    ]);
-
-    const peerBuyOrdersByPairs = groupBy(peerOrders.buyOrders, order => order.pairId);
-    const peerSellOrdersByPairs = groupBy(peerOrders.sellOrders, order => order.pairId);
-    const ownBuyOrdersByPairs = groupBy(ownOrders.buyOrders, order => order.pairId);
-    const ownSellOrdersByPairs = groupBy(ownOrders.sellOrders, order => order.pairId);
+  public init = async () => {
+    const pairs = await this.repository.getPairs();
 
     pairs.forEach((pair) => {
       this.matchingEngines[pair.id] = new MatchingEngine(
         pair.id,
         this.config.internalmatching,
-        peerBuyOrdersByPairs[pair.id],
-        peerSellOrdersByPairs[pair.id],
-        ownBuyOrdersByPairs[pair.id],
-        ownSellOrdersByPairs[pair.id],
       );
+      this.ownOrders[pair.id] = this.initOrders();
+      this.peerOrders[pair.id] = this.initOrders();
     });
 
     this.pairs = pairs;
   }
 
+  private initOrders = (): Orders => {
+    return {
+      buyOrders: [],
+      sellOrders: [],
+    };
+  }
+
   /**
    * Returns the list of available trading pairs.
    */
-  public getPairs(): Promise<db.PairInstance[]> {
+  public getPairs = (): Promise<db.PairInstance[]> => {
     return this.repository.getPairs();
   }
 
   /**
    * Returns lists of buy and sell orders of peers sorted by price.
    */
-  public getPeerOrders(pairId: string, maxResults: number): Promise<Orders> {
-    return this.repository.getPeerOrders(pairId, maxResults);
+  public getPeerOrders = (pairId: string, maxResults: number): Orders => {
+    return this.getOrders(maxResults, this.peerOrders[pairId]);
   }
 
   /*
   * Returns lists of the node's own buy and sell orders sorted by price.
   */
-  public getOwnOrders(pairId: string, maxResults: number): Promise<Orders> {
-    return this.repository.getOwnOrders(pairId, maxResults);
+  public getOwnOrders = (pairId: string, maxResults: number): Orders => {
+    return this.getOrders(maxResults, this.ownOrders[pairId]);
+  }
+
+  private getOrders = (maxResults: number, orders: Orders) => {
+    if (maxResults > 0) {
+      const result = this.initOrders();
```

Comment:
any reason for still doing a push into a result object, and not merely returning an object with the slice result?

Reply:
I was referring to doing something like this:
```
return {
  buyOrders: orders.buyOrders.slice(…
  sellOrders: orders.sellOrders.slice(…
}
```

Nothing major, just thought it can make the code even more simple.